## Практическое задание

<ol>
    <li>Попробуйте обучить нейронную сеть на Keras с другими параметрами. 
        Опишите в комментарии к уроку - какой результата вы добились от нейросети? Что помогло вам улучшить ее точность?</li>
    <li>Поработайте с документацией Keras.</li>
</ol>

In [5]:
# The full neural network code!
###############################
import numpy as np
import mnist
from keras.models import Sequential
from keras.layers import Dense
from keras.utils import to_categorical

In [132]:
train_images = mnist.train_images()
train_labels = mnist.train_labels()
test_images = mnist.test_images()
test_labels = mnist.test_labels()

# Normalize the images.
train_images = (train_images / 255) - 0.5
test_images = (test_images / 255) - 0.5


# Flatten the images.
train_images = train_images.reshape((-1, 784))
test_images = test_images.reshape((-1, 784))

In [148]:
# Build the model.
model = Sequential([
  Dense(64, activation='elu', input_shape=(784,)),
  Dense(64, activation='elu'),
    Dense(64, activation='elu'),
#         Dense(64, activation='relu'),
  Dense(10, activation='softmax'),
])

In [154]:
# Build the model.
model = Sequential([
  Dense(64, activation='relu', input_shape=(784,)),
  Dense(128, activation='relu'),
    Dense(64, activation='relu'),
  Dense(10, activation='softmax'),
])

In [155]:

# Compile the model.
model.compile(
  optimizer='Adadelta',
  loss='cosine_similarity',
  metrics=['top_k_categorical_accuracy'],
)

In [151]:

# Compile the model.
model.compile(
  optimizer='Adam',
  loss='cosine_similarity',
  metrics=['top_k_categorical_accuracy'],
)

In [143]:


# Train the model.
model.fit(
  train_images,
  to_categorical(train_labels),
  epochs=5,
  batch_size=32,
)

# Evaluate the model.
model.evaluate(
  test_images,
  to_categorical(test_labels)
)

# Save the model to disk.
model.save_weights('model1.h5')

# Load the model from disk later using:
# model.load_weights('model.h5')


Epoch 1/5
60000/60000 [==============================] - 3s 55us/step - loss: -0.8964 - top_k_categorical_accuracy: 0.9881
Epoch 2/5
60000/60000 [==============================] - 3s 50us/step - loss: -0.9469 - top_k_categorical_accuracy: 0.9966
Epoch 3/5
60000/60000 [==============================] - 3s 50us/step - loss: -0.9595 - top_k_categorical_accuracy: 0.9974
Epoch 4/5
60000/60000 [==============================] - 3s 51us/step - loss: -0.9657 - top_k_categorical_accuracy: 0.9980
Epoch 5/5
10000/10000 [==============================] - 0s 29us/step


In [152]:


# Train the model.
model.fit(
  train_images,
  to_categorical(train_labels),
  epochs=5,
  batch_size=32,
)

# Evaluate the model.
model.evaluate(
  test_images,
  to_categorical(test_labels)
)

# Save the model to disk.
model.save_weights('model2.h5')

# Load the model from disk later using:
# model.load_weights('model.h5')


Epoch 1/5
60000/60000 [==============================] - 3s 51us/step - loss: -0.9069 - top_k_categorical_accuracy: 0.9901
Epoch 2/5
60000/60000 [==============================] - 3s 48us/step - loss: -0.9466 - top_k_categorical_accuracy: 0.9969
Epoch 3/5
60000/60000 [==============================] - 3s 48us/step - loss: -0.9565 - top_k_categorical_accuracy: 0.9977
Epoch 4/5
60000/60000 [==============================] - 3s 48us/step - loss: -0.9610 - top_k_categorical_accuracy: 0.9980
Epoch 5/5
10000/10000 [==============================] - 0s 21us/step


In [156]:


# Train the model.
model.fit(
  train_images,
  to_categorical(train_labels),
  epochs=5,
  batch_size=32,
)

# Evaluate the model.
model.evaluate(
  test_images,
  to_categorical(test_labels)
)

# Save the model to disk.
model.save_weights('model3.h5')

# Load the model from disk later using:
# model.load_weights('model.h5')


Epoch 1/5
60000/60000 [==============================] - 3s 58us/step - loss: -0.8977 - top_k_categorical_accuracy: 0.9869
Epoch 2/5
60000/60000 [==============================] - 3s 53us/step - loss: -0.9494 - top_k_categorical_accuracy: 0.9966
Epoch 3/5
60000/60000 [==============================] - 3s 55us/step - loss: -0.9603 - top_k_categorical_accuracy: 0.9976
Epoch 4/5
60000/60000 [==============================] - 3s 55us/step - loss: -0.9657 - top_k_categorical_accuracy: 0.9983
Epoch 5/5
10000/10000 [==============================] - 0s 21us/step


In [157]:

# Predict on the first 10 test images.
predictions = model.predict(test_images[:10])

# Print our model's predictions.
print(np.argmax(predictions, axis=1)) # [7, 2, 1, 0, 4]

# Check our predictions against the ground truths.
print(test_labels[:10]) # [7, 2, 1, 0, 4]

[7 2 1 0 4 1 4 9 5 9]
[7 2 1 0 4 1 4 9 5 9]


## Эксперименты и выводы
Если убрать из нормализации центровку на 0, то качество растёт на 1-2%
Попробовал заменить при этом relu на softplus - метрики упали; на tanh - метрики вернулись.

При сокращении до двух слоёв качество падает. При этом, если поставить на вход softmax, начинаются ошибки. Точость около 0.7. Увеличение слоёв не даёт значительного прироста качества.

Заменил оптимайзер на Adadelta, после чего сразу получил более 98% по всем эпохам. Вероятно, из-за того, что не перегрузил данные.
Nadam даёт 97% ; SGD - низкий результат 92% и ошибки.

Метрика top_k_categorical_accuracy дала прирост до 99% по всем эпохам. categorical_accuracy - вернула к значениям вокруг 95%
При оптимайзере Adam показатели также высокие (99%), но есть ошибки, что говорит о переобучении.

### Вывод:
Усложнение сети не даёт значимого прироста. Но и не влияет на производительность на простом примере.
Максимальный эффект дают настройки модели  
   * optimizer='Adadelta',
   * loss='cosine_similarity',
   * metrics='top_k_categorical_accuracy'


Использовались параметры из документации: 
https://keras.io/api/losses/
https://keras.io/api/optimizers/
https://keras.io/api/metrics/